In [1]:
# ==========================================================
# IMPORTAÇÃO DAS BIBLIOTECAS
# ==========================================================

import pandas as pd

# Divide os dados em treino e teste
from sklearn.model_selection import train_test_split

# Faz a normalização dos dados
from sklearn.preprocessing import MinMaxScaler

# Algoritmo KNN pronto da biblioteca scikit-learn
from sklearn.neighbors import KNeighborsClassifier

# Métricas de avaliação
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [58]:
# ==========================================================
# LEITURA DO DATASET
# ==========================================================

df = pd.read_csv("https://drive.google.com/uc?export=download&id=1wF7amDwgNKIxa9Lc5Dm4OXkiDy_N_CYu")

print("Formato do dataset:", df.shape)
print(df.head())

Formato do dataset: (1998, 20)
   battery_power  bluetooth  clock_speed  dual_sim  front_camera  four_g  \
0            667          1          0.6         0             6       1   
1           1162          0          2.0         0             0       1   
2           1712          1          0.9         1             2       0   
3           1696          0          1.7         0             1       0   
4            592          0          0.8         1             1       0   

   storage  mobile_weight  n_cores  primary_camera  px_height  px_width   ram  \
0        3            185        6              13         13      1463  2855   
1       56            114        7               1         56       997  1652   
2       23            155        7              19        155       663  1422   
3       13             97        3               2        179       772   523   
4       52            103        7              10        332       970  1507   

   sc_height  sc_width  t

In [101]:
# ==========================================================
# ESCOLHA DOS ATRIBUTOS
# ==========================================================

# Mesmas 8 colunas utilizadas no KNN Manual
colunas_usadas = [
    "battery_power",
    "clock_speed",
    "storage",
    "n_cores",
    "primary_camera",
    "px_width",
    "ram",
    "talk_time"
]

# X = atributos usados para prever
X = df[colunas_usadas]

# y = variável alvo
y = df["price_range"]

In [102]:
# ==========================================================
# DIVISÃO ENTRE TREINO E TESTE
# ==========================================================

# 80% treino e 20% teste
# random_state mantém o mesmo sorteio sempre que rodar
# stratify=y mantém a proporção das classes em treino e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Quantidade treino:", len(X_treino))
print("Quantidade teste:", len(X_teste))

Quantidade treino: 1598
Quantidade teste: 400


In [103]:
# ==========================================================
# NORMALIZAÇÃO DOS DADOS
# ==========================================================

# Cria o normalizador Min-Max
normalizador = MinMaxScaler()

# Aprende os mínimos e máximos usando apenas os dados de treino
X_treino_normalizado = normalizador.fit_transform(X_treino)

# Aplica a mesma normalização nos dados de teste
X_teste_normalizado = normalizador.transform(X_teste)

print("Dados normalizados com sucesso!")

Dados normalizados com sucesso!


In [132]:
# ==========================================================
# CRIAÇÃO DO MODELO KNN
# ==========================================================

# Define o valor de K
k = 25

# Cria o modelo KNN
modelo_knn = KNeighborsClassifier(n_neighbors=k)

print("Valor escolhido para K:", k)

Valor escolhido para K: 25


In [133]:
# ==========================================================
# TREINAMENTO DO MODELO
# ==========================================================

# O modelo aprende usando os dados de treino
modelo_knn.fit(X_treino_normalizado, y_treino)

KNeighborsClassifier(n_neighbors=25)

In [134]:
# ==========================================================
# PREVISÕES
# ==========================================================

# O modelo faz previsões nos dados de teste
previsoes = modelo_knn.predict(X_teste_normalizado)

In [135]:
# ==========================================================
# AVALIAÇÃO DO MODELO
# ==========================================================

# Calcula a acurácia
acuracia = accuracy_score(y_teste, previsoes)

print("\n================ RESULTADOS ================")
print("Total de testes:", len(y_teste))
print("Acurácia:", acuracia)
print("Acurácia em porcentagem:", round(acuracia * 100, 2), "%")


================ RESULTADOS ================
Total de testes: 400
Acurácia: 0.7225
Acurácia em porcentagem: 72.25 %


In [136]:
# ==========================================================
# MATRIZ DE CONFUSÃO
# ==========================================================

print("\nMatriz de confusão:")
print(confusion_matrix(y_teste, previsoes))


Matriz de confusão:
[[88 12  0  0]
 [17 59 24  0]
 [ 0 27 60 13]
 [ 0  0 18 82]]


In [137]:
# ==========================================================
# RELATÓRIO DE CLASSIFICAÇÃO
# ==========================================================

print("\nRelatório de classificação:")
print(classification_report(y_teste, previsoes))


Relatório de classificação:
              precision    recall  f1-score   support

           0       0.84      0.88      0.86       100
           1       0.60      0.59      0.60       100
           2       0.59      0.60      0.59       100
           3       0.86      0.82      0.84       100

    accuracy                           0.72       400
   macro avg       0.72      0.72      0.72       400
weighted avg       0.72      0.72      0.72       400



### Usando todas as colunas para treino!   


### RESULTADOS COM K=1
Acurácia em porcentagem: 35.5 % / f1-score 0.36

### RESULTADOS COM K=9
Acurácia em porcentagem: 45.5 % / f1-score 0.46    

### RESULTADOS COM K=15
Acurácia em porcentagem: 48 % / f1-score 0.48   

### RESULTADOS COM K=21
Acurácia em porcentagem: 48.75 % / f1-score 0.49   

### RESULTADOS COM K=23 **************
Acurácia em porcentagem: 49.25 % / f1-score 0.49

### RESULTADOS COM K=25
Acurácia em porcentagem: 48 % / f1-score 0.48






### Usando 14 colunas para treino!   


### RESULTADOS COM K=1
Acurácia em porcentagem: 51.75.5 % / f1-score 0.53

### RESULTADOS COM K=9 ****************
Acurácia em porcentagem: 57.25 % / f1-score 0.57    

### RESULTADOS COM K=11
Acurácia em porcentagem: 56.75 % / f1-score 0.57   

### RESULTADOS COM K=15
Acurácia em porcentagem: 55 % / f1-score 0.55   








### Usando 8 colunas para treino!   


### RESULTADOS COM K=1
Acurácia em porcentagem: 64 % / f1-score 0.64

### RESULTADOS COM K=9
Acurácia em porcentagem: 67.25 % / f1-score 0.68    

### RESULTADOS COM K=15
Acurácia em porcentagem: 71.25 % / f1-score 0.72   

### RESULTADOS COM K=19 ****************
Acurácia em porcentagem: 73.75 % / f1-score 0.74   

### RESULTADOS COM K=25
Acurácia em porcentagem: 72.25 % / f1-score 0.72   








Colunas Usadas no treino de 14:

    battery_power",
    "clock_speed",
    "storage",
    "mobile_weight",
    "n_cores",
    "primary_camera",
    "px_height",
    "px_width",
    "ram",
    "sc_height",
    "sc_width",
    "talk_time",
    "four_g",
    "wifi"

Colunas Usadas no treino de 8:

    "battery_power",
    "clock_speed",
    "storage",
    "n_cores",
    "primary_camera",
    "px_width",
    "ram",
    "talk_time"

